test/e2e_node/device_plugin_failures_test.go

```diff
@@ -0,0 +1,247 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2enode
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+
+	kubeletdevicepluginv1beta1 "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e_node/testdeviceplugin"
+)
+
+// Serial because the test restarts Kubelet
+var _ = SIGDescribe("Device Plugin Failures:", framework.WithNodeConformance(), func() {
+	f := framework.NewDefaultFramework("device-plugin-failures")
+
+	var getNodeAllocatableValue = func(resourceName string) int {
+		node, err := f.ClientSet.CoreV1().Nodes().Get(context.Background(), framework.TestContext.NodeName, metav1.GetOptions{})
+
+		// it is a test failure if node status is not available
+		gomega.Expect(err).To(gomega.Succeed())
+
+		for key, val := range node.Status.Allocatable {
+			resource := string(key)
+			if resource == resourceName {
+				return int(val.Value())
+			}
+		}
+
+		return -1
+	}
+
+	ginkgo.It("when GetDevicePluginOptions fails, device plugin will not be used", func(ctx context.Context) {
+		// randomizing so tests can run in parallel
+		resourceName := fmt.Sprintf("test.device/%s", f.UniqueName)
+
+		expectedErr := fmt.Errorf("GetDevicePluginOptions failed")
+
+		plugin := testdeviceplugin.NewDevicePlugin(func(name string) error {
+			if name == "GetDevicePluginOptions" {
+				return expectedErr
+			}
+			return nil
+		})
+
+		err := plugin.RegisterDevicePlugin(f.UniqueName, resourceName, []kubeletdevicepluginv1beta1.Device{{ID: "testdevice", Health: kubeletdevicepluginv1beta1.Healthy}})
+		defer plugin.Stop() // should stop even if registration failed
+		gomega.Expect(err).To(gomega.MatchError(gomega.ContainSubstring("failed to get device plugin options")))
+		gomega.Expect(err).To(gomega.MatchError(gomega.ContainSubstring(expectedErr.Error())))
```

Comment:
Can we move towards well known errors and use `errors.Is`/`errors.As`?

Reply:
I think the error string is not public in the plugin